baseapp/baseapp.go

```diff
@@ -497,7 +503,10 @@ func (app *BaseApp) setState(mode execMode, h cmtproto.Header) {
 
 	switch mode {
 	case execModeCheck:
-		baseState.SetContext(baseState.Context().WithIsCheckTx(true).WithMinGasPrices(app.minGasPrices))
+		// should take a snapshot of the multistore and cache it
+		// to prevent state corruption during checktx or simulate.
+		baseState.ms, _ = app.cms.CacheMultiStoreWithVersion(app.cms.LatestVersion())
```

Comment:
I can see how this may solve your collusions but when simulations are executed on "older" state, this may lead to invalid TX that are be rejected by the chain.
For example, re-using the nonce of a pending TX would work in simulations.

Reply:
I don't think this is "older" state. 

This function will be called after app.cms.Commit(), so the app.cms.LatestVersion() is the latest state not "older" state.

after this, comet will trigger recheck and this recheck operation will update the checkState to apply pending txs ante state updates. 

so no nonce problem we have now.

Reply:
setState for CheckTx is being called only on InitChain and Commit. Meaning that when you read off checkTx's state you should be reading from the last block. It won't include anything from the current block, as that's being written in finalizeBlock's state.

Reply:
yea I mean every Commit will update the latest state(cms) and that latest state will be `checkTxState`. Is this right?

then, every (re)check tx ante will update this checkTxState.